### ISSUE TITLE
Consider leveraging `Canvas::captureMedia` for generating content

### ISSUE THREAD

**Issue by mhofman** [w3c/mediacapture-worker#31]
In issue #30, I mentioned we might be able to leverage the `Canvas::captureMedia` feature to generate a `MediaStreamTrack` after processing is done in an `OffscreenCanvas` fed by the monitoring part of this spec.

The issue of keeping A/V sync was raised, since the `captureMedia` API doesn't allow the input of frame timing.

I think it would be great if we could reconcile this feature gap and somehow avoid duplicated features for generating video tracks.

**Comment by mhofman** [comment]
Continuing discussion from comments in #30:

> > A MediaStream actually doesn't guarantee that its tracks are played perfectly in sync, just that they're related.
> 
> I think it should. It certainly does in Gecko!

According to the spec:

> All tracks in a MediaStream are intended to be synchronized when rendered. This is not a hard requirement, since it might not be possible to synchronize tracks from sources that have different clocks.

So we're both right, it's not a guarantee, but it's a best effort when possible. Interestingly, the spec mentions the clock issue. Processing Video and Audio independently causes their clocks to not be shared anymore. This spec proposal seem to work around that by "manually syncing" the clocks.

> > How do the audio specs handle the manual timing aspect?
> 
> By exposing time and sample rates explicitly. A sample rate isn't very suitable for video though, because of dropped frames and video formats that support irregular frame timing.

Right so there is nothing in the "input" of the audio generation that's specific to timing, just that the destination has an intrinsic sampling rate, so the timing comes from the amount of audio samples that are generated in the input.

As you mention that model doesn't really fit video data when framerate is non-constant.
I'm not sure how to cleanly solve this discrepancy, but it sure would be nice to have a consistent model.

**Comment by mhofman** [comment]
Looking more at the WebAudio spec, it seem to actually solve the issue of offline processing separately form MediaStreams, which are kept real-time: the `OfflineAudioContext` generates an `AudioBuffer` and there doesn't seem to be a way to generate a `MediaStreamTrack` based on offline audio.
There is a `MediaStreamAudioSourceNode`, but that's a regular `AudioNode` for real-time data.

**Comment by kakukogou** [comment]
> the OfflineAudioContext generates an AudioBuffer and there doesn't seem to be a way to generate a MediaStreamTrack based on offline audio.

There is `MediaStreamAudioDestinationNode`. By connecting an `AudioBufferSourceNode` to a `MediaStreamAudioDestinationNode`, you can generate a real-time `MediaStream` from an offline `AudioBuffer`.

**Comment by mhofman** [comment]
From what I understand you cannot attach that node to a graph backed by an offline context.

**Comment by kakukogou** [comment]
Yes, we can't. The scenario of connecting an `AudioBufferSourceNode` to a `MediaStreamAudioDestinationNode` can only happens within the normal `AudioContext`.
